stubgen: Fix call-based namedtuple omitted from class bases

hamdanal opened the pull request:
Fixes #9901
Fixes #13662

Fix inheriting from a call-based `collections.namedtuple` / `typing.NamedTuple` definition that was omitted from the generated stub.

This automatically adds support for the call-based `NamedTuple` in general not only in class bases (Closes #13788).

<details>
<summary>An example before and after</summary>
Input:

```python
import collections
import typing
from collections import namedtuple
from typing import NamedTuple

CollectionsCall = namedtuple("CollectionsCall", ["x", "y"])

class CollectionsClass(namedtuple("CollectionsClass", ["x", "y"])):
    def f(self, a):
        pass

class CollectionsDotClass(collections.namedtuple("CollectionsClass", ["x", "y"])):
    def f(self, a):
        pass

TypingCall = NamedTuple("TypingCall", [("x", int | None), ("y", int)])

class TypingClass(NamedTuple):
    x: int | None
    y: str

    def f(self, a):
        pass

class TypingClassWeird(NamedTuple("TypingClassWeird", [("x", int | None), ("y", str)])):
    z: float | None

    def f(self, a):
        pass

class TypingDotClassWeird(typing.NamedTuple("TypingClassWeird", [("x", int | None), ("y", str)])):
    def f(self, a):
        pass
```

Output diff (before and after):
```diff
diff --git a/before.pyi b/after.pyi
index c88530e2c..95ef843b4 100644
--- a/before.pyi
+++ b/after.pyi
@@ -1,26 +1,29 @@
+import typing
 from _typeshed import Incomplete
 from typing_extensions import NamedTuple

 class CollectionsCall(NamedTuple):
     x: Incomplete
     y: Incomplete

-class CollectionsClass:
+class CollectionsClass(NamedTuple('CollectionsClass', [('x', Incomplete), ('y', Incomplete)])):
     def f(self, a) -> None: ...

-class CollectionsDotClass:
+class CollectionsDotClass(NamedTuple('CollectionsClass', [('x', Incomplete), ('y', Incomplete)])):
     def f(self, a) -> None: ...

-TypingCall: Incomplete
+class TypingCall(NamedTuple):
+    x: int | None
+    y: int

 class TypingClass(NamedTuple):
     x: int | None
     y: str
     def f(self, a) -> None: ...

-class TypingClassWeird:
+class TypingClassWeird(NamedTuple('TypingClassWeird', [('x', int | None), ('y', str)])):
     z: float | None
     def f(self, a) -> None: ...

-class TypingDotClassWeird:
+class TypingDotClassWeird(typing.NamedTuple('TypingClassWeird', [('x', int | None), ('y', str)])):
     def f(self, a) -> None: ...
```
</details>

ikonst commented on mypy/stubgen.py:
```diff
@@ -1038,36 +1076,81 @@ def visit_assignment_stmt(self, o: AssignmentStmt) -> None:
+            return [(item, "Incomplete") for item in items]
+        elif self.is_typed_namedtuple(call):
+            fields: list[tuple[str, str]] = []
+            if isinstance(call.args[1], (ListExpr, TupleExpr)):
```
Consider early return

ikonst commented on mypy/stubgen.py:
```diff
@@ -1038,36 +1076,81 @@ def visit_assignment_stmt(self, o: AssignmentStmt) -> None:
+                self.import_tracker.require_name("Incomplete")
+            return [(item, "Incomplete") for item in items]
+        elif self.is_typed_namedtuple(call):
+            fields: list[tuple[str, str]] = []
```
Consider moving after `if`

ikonst commented on mypy/stubgen.py:
```diff
@@ -986,6 +992,35 @@ def get_base_types(self, cdef: ClassDef) -> list[str]:
+            elif isinstance(base, CallExpr):
+                # namedtuple(typename, fields), NamedTuple(typename, fields) calls can
+                # be used as a base class. The first argument is a string literal that
+                # is usually the same as the class name.
```
```suggestion
                # is usually the same as the class name.
                #
```

ikonst commented on mypy/stubgen.py:
```diff
@@ -986,6 +992,35 @@ def get_base_types(self, cdef: ClassDef) -> list[str]:
+                # namedtuple fields at runtime.
+                if self.is_namedtuple(base):
+                    nt_fields = self._get_namedtuple_fields(base)
+                    typename = cast(StrExpr, base.args[0]).value
```
consider isinstance assertion

hamdanal commented:
I made the requested changes.
I also deleted the `NamedTuplePrinter` class that I added in an earlier commit and moved the logic to the existing `AliasPrinter`. This makes the code a bit cleaner.

ikonst commented:
Curious how this change didn't invalidate a test
```python
-and f"{callee.expr.name}.{callee.name}" in "collections.namedtuple"
+and f"{callee.expr.name}.{callee.name}" == "collections.namedtuple"
```

hamdanal commented:
> Curious how this change didn't invalidate a test
> 
> ```python
> -and f"{callee.expr.name}.{callee.name}" in "collections.namedtuple"
> +and f"{callee.expr.name}.{callee.name}" == "collections.namedtuple"
> ```

I think because `"collections.namedtuple" in "collections.namedtuple"` is True

ikonst commented on mypy/stubgen.py:
```diff
@@ -1038,36 +1075,82 @@ def visit_assignment_stmt(self, o: AssignmentStmt) -> None:
+                self.import_tracker.require_name("Incomplete")
+            return [(item, "Incomplete") for item in items]
+        elif self.is_typed_namedtuple(call):
+
+            if isinstance(call.args[1], (ListExpr, TupleExpr)):
+                fields: list[tuple[str, str]] = []
+                b = AliasPrinter(self)
+                for item in call.args[1].items:
+                    if isinstance(item, TupleExpr) and len(item.items) == 2:
+                        field_name, field_type = item.items
+                        if not isinstance(field_name, StrExpr):
+                            return None  # Invalid NamedTuple field name
+                        fields.append((field_name.value, field_type.accept(b)))
+                    else:
+                        return None  # Invalid NamedTuple field type
+                return fields
+            else:
+                return None  # Invalid NamedTuple fields type
```
With early returns:
```suggestion
            fields_arg = call.args[1]
            if not isinstance(fields_arg, (ListExpr, TupleExpr)):
                return None

            fields: list[tuple[str, str]] = []
            b = AliasPrinter(self)
            for item in fields_arg.items:
                if not isinstance(item, TupleExpr) and len(item.items) == 2:
                    return None  # Invalid NamedTuple field type
                field_name, field_type = item.items
                if not isinstance(field_name, StrExpr):
                    return None  # Invalid NamedTuple field name
                fields.append((field_name.value, field_type.accept(b)))
            return fields
```

ikonst replied:
BTW, on failure to parse, do we want to emit "Incomplete" or something rather than just ignore?

hamdanal replied:
Done. I also made the same change in the section before

hamdanal replied:
> BTW, on failure to parse, do we want to emit "Incomplete" or something rather than just ignore?

We emit Incomplete. It is done in the `process_namedtuple` method (this is the original behavior since before this PR).

ikonst commented:
> I think because "collections.namedtuple" in "collections.namedtuple" is True

🤦 🤣

ikonst commented:
(note that I'm not a maintainer, just a fellow contributor)

AlexWaygood commented on test-data/unit/stubgen.test:
```diff
@@ -692,13 +710,60 @@ class Z(NamedTuple):
+from _typeshed import Incomplete
+from typing import NamedTuple
+
+class X(NamedTuple('X', [('a', Incomplete), ('b', Incomplete)])): ...
```
Hmm, in typeshed we prefer to avoid call-based namedtuples wherever possible, so we would usually do something like this for this kind of thing:

```py
class _XBase(NamedTuple):
    a: Incomplete
    b: Incomplete

class X(_XBase): ...
```

I'm guessing it might be hard to achieve that from stubgen, though?

hamdanal replied:
I agree, call-based syntax isn't desirable. It is however tricky to get the conversion to a stub-only class definition right. In fact I tried briefly when I started with the PR then decided it is safer and much simpler to keep the call-based definition.

Note that even though this syntax is ugly/undesirable, it type-checks perfectly fine by both mypy and pyright and this PR still fixes the issue where stubgen generated wrong stubs without any warning. It also makes a posterior step of manual conversion to a class definition much simpler as the information is now there in the stub instead of having to grep the python source for all namedtuple bases in class definitions.

Having said that, I don't mind implementing this, whether in this PR or in a follow up one. I do like to get the opinion of a stubgen maintainer/expert before working on it though as it will require some work. If the answer is do it, I have a couple of questions regarding this step that I can ask then.

hamdanal replied:
@AlexWaygood are you comfortable enough this idea will be welcome so that I can work on it or should we ping someone?

JelleZijlstra replied:
I think it's fine to generate the stubs as in this PR. If you fix the merge conflict, I can review this PR.

JelleZijlstra commented on mypy/stubgen.py:
```diff
@@ -1010,6 +1012,37 @@ def get_base_types(self, cdef: ClassDef) -> list[str]:
+                        # Incomplete as field types
+                        fields_str = ", ".join(f"({f!r}, {t})" for f, t in nt_fields)
+                    else:
+                        # Invalid namedtuple() call, cannot determine fields
```
Looks like you generate `NamedTuple(name, [Incomplete])` as the base in this case. That seems wrong, as it's an invalid type. I think we should make `Incomplete` the base in this case.

JelleZijlstra replied:
Also, is this code path tested?

hamdanal replied:
Good call. I fixed it. The tests now cover the invalid namedtuple call code path.
